Add 0036-trusted-artifacts (#195)
konflux-ci/architecture: ADR/0036-trusted-artifacts.md

# 36. Trusted Artifacts

Date: 2024-06-11

## Status

Accepted

## Context

One of the properties of Konflux is that users should be allowed to include their own Tekton Tasks

> in a build Pipeline, e.g. to execute unit tests, without invalidating the build process. This is

Reviewer: Perhaps

```suggestion
in a build Pipeline, e.g. to execute unit tests, without jeopardizing the integrity of the build process. This is
```

Author: Better. Thanks.

distinct from other build systems where a rigid process prevents users from applying such
customizations. To support this, Konflux build Pipelines use Trusted Artifacts to securely share

> data between Tasks. Enterprise Contract is then responsible for verifying that Trusted Artifacts

Reviewer: ```suggestion
files between Tasks. Enterprise Contract is then responsible for verifying that Trusted Artifacts
```

Author: Good to be specific. Also modified other occurrences.

were properly used in the parts of the build Pipeline that affect the build outcome, e.g.
`git-clone` and `buildah` Tasks.

> Trusted Artifacts is inspired by the upcoming work being done by the Tekton Community,

Author: The upstream work will require changes to the Tekton resources IIUC. I don't think there's a way to avoid a migration. But hopefully, this is isolated to Task definitions and we can switch over without requiring a migration from the user side.

[TEP-0139](https://github.com/tektoncd/community/blob/main/teps/0139-trusted-artifacts.md). The
Konflux version is meant to be a stop-gap until that feature is implemented and ready to be used.
When the time comes, the Konflux implementation should align with what is provided by the Tekton

> Community and this ADR may need a revision.

Reviewer 2: A brief summary of what make trusted artifacts "trusted" and how do they enable pipeline security would be welcome here.

All I could gather from this text is that they are stored in an OCI registry...

Author: Excellent idea! Added a paragraph. Let me know what you think

Reviewer 2: looks good, what is preventing tampering in the case? The fact that the checksum is passed via channel that is internal to the pipeline?

applying such customizations. To support this, Konflux build Pipelines use Trusted Artifacts to
securely share files between Tasks. Enterprise Contract is then responsible for verifying that
Trusted Artifacts were properly used in the parts of the build Pipeline that affect the build
outcome, e.g. `git-clone` and `buildah` Tasks.

Trusted Artifacts is inspired by the upcoming work being done by the Tekton Community,
[TEP-0139](https://github.com/tektoncd/community/blob/main/teps/0139-trusted-artifacts.md). The
Konflux version is meant to be a stop-gap until that feature is implemented and ready to be used.
When the time comes, the Konflux implementation should align with what is provided by the Tekton
Community and this ADR may need a revision.

In brief, the processes of *creating* a Trusted Artifact wraps files into an archive. Then, the
location of the archive and its checksum digests are recorded as a Task result. The process of
*consuming* a Trusted Artifact extracts such archive, while verying its checksum digest, into a
volume only accessible to the Task, e.g. `emptyDir`. The name and checksum digest of the archive is

> provided via Task parameters.

Reviewer 2: ```suggestion
provided via Task parameters.

With the artifact checksums being passed over the task parameters we can ensure that the artifacts are bot being tampered with as they are passed between different pipeline tasks.
```

Trying to summarize our other thread

Author: TY! I reworded this a bit. Let me know what you think

# 36. Trusted Artifacts

Date: 2024-06-11

## Status

Accepted

## Context

One of the properties of Konflux is that users should be allowed to include their own Tekton Tasks
in a build Pipeline, e.g. to execute unit tests, without jeopardizing the integrity of the build
process. This is distinct from other build systems where a rigid process prevents users from
applying such customizations. To support this, Konflux build Pipelines use Trusted Artifacts to
securely share files between Tasks. Enterprise Contract is then responsible for verifying that

> Trusted Artifacts were properly used in the parts of the build Pipeline that affect the build

Reviewer 3: It also is used to verify that the contents were not tainted by untrusted tasks.

Author: I think this falls under the "Trusted Artifacts were properly used" comment.

outcome, typically the `git-clone`, `prefetch-dependencies`, and `buildah` Tasks.

Trusted Artifacts is inspired by the upcoming work being done by the Tekton Community,
[TEP-0139](https://github.com/tektoncd/community/blob/main/teps/0139-trusted-artifacts.md). The
Konflux version is meant to be a stop-gap until that feature is implemented and ready to be used.
When the time comes, the Konflux implementation should align with what is provided by the Tekton

> Community and this ADR may need a revision.

Reviewer 3: Can we state that a new ADR should be created to represent the fact that we are changing as well?

Author: Good idea. Done.

> In brief, the processes of *creating* a Trusted Artifact wraps files into an archive. Then, the

Reviewer 3: Do we need to be more specific about what the archive is? These archives also need to be tamper-proof.

Author: I really don't want this ADR to be a design doc of Trusted Artifacts. I think that belongs somewhere else. https://github.com/konflux-ci/build-trusted-artifacts is likely a good place for it.

location of the archive and its checksum digests are recorded as a Task result. The process of

> *consuming* a Trusted Artifact extracts such archive, while verifying its checksum digest, into a

Reviewer 4: ```suggestion
*consuming* a Trusted Artifact extracts such an archive, while verifying its checksum digest, into a
```

volume only accessible to the Task, e.g. `emptyDir`. The name and the checksum digest of the archive
is provided via Task parameters. This ensures the artifacts produced by one Task are not tampered
with when they are consumed by other Tasks.

Furthermore, Konflux takes the approach of sharing such artifacts between Tasks via an OCI registry,
e.g. quay.io, instead of using Tekton Workspaces backed by Persistent Volume Claims. This has
several advantages that were previously discussed
[here](https://github.com/konflux-ci/build-definitions/pull/913#issue-2215784386).

## Decision

> Sharing files between Tasks is done via Trusted Artifacts backed by OCI storage.

Reviewer 3: Do we want to state that we cannot use the referrer's api here because these archives are generated before we produce an image? I don't know if anyone might wonder about it. (no is a fine response).

Author: I think that's too low level for this doc.
